src/java.base/macosx/classes/sun/nio/fs/MacOSXFileSystem.java

```diff
@@ -35,6 +35,12 @@
 
 class MacOSXFileSystem extends BsdFileSystem {
 
+    private static final String PROPERTY_ENABLE_FILE_NAME_ENCODING =
+        "java.nio.file.Path.ENABLE_FILE_NAME_ENCODING";
+
+    private static final boolean ENCODE_FILE_NAMES =
+        Boolean.getBoolean(PROPERTY_ENABLE_FILE_NAME_ENCODING);
```

Comment:
APFS is the default since macOS 10.13 but I agree, it's safer to have a system property for a few releases in case anything is relying on this.

I assume you'll need to use GetPropertyAction.privilegedGetProperty so that it works when there is a security manager set.

As regards the property name then it will need to jdk.something because it's JDK specific rather than a standard property. Maybe something like -Djdk.nio.path.useNormalizationFormD ?

Reply:
Fixed in ab2e42dbfc47ffe9026e769ef5d9329ee9098452.